dali/operators/reader/gds_mem.cc

```diff
@@ -12,40 +12,283 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#include <map>
 #include <memory>
+#include <mutex>
 #include <utility>
 #include <vector>
 #include "dali/operators/reader/gds_mem.h"
+#include "dali/util/cufile_helper.h"
+#include "dali/core/math_util.h"
 #include "dali/core/spinlock.h"
 #include "dali/core/mm/pool_resource.h"
+#include "dali/core/mm/detail/align.h"
 #include "dali/core/mm/malloc_resource.h"
 #include "dali/core/mm/composite_resource.h"
 
 namespace dali {
+namespace gds {
+
+static size_t GetGDSChunkSizeEnv() {
+  if (char *env = getenv("DALI_GDS_CHUNK_SIZE")) {
+    size_t s = atoll(env);
+    DALI_ENFORCE(!is_pow2(s), "GDS chunk size must be a power of 2");
+    DALI_ENFORCE(s >= kGDSAlignment && s <= (16 << 20),
+      make_string("GDS chunk size must be a power of two between ", kGDSAlignment, " and 16 M"));
+    return s;
+  } else {
+    return 2 << 20;  // default - 2 MiB
+  }
+}
+
+size_t GetGDSChunkSize() {
+  static size_t chunk_size = GetGDSChunkSizeEnv();
+  return chunk_size;
+}
+
+void RegisterChunks(void *start, int chunks, size_t chunk_size) {
+  char *addr = static_cast<char *>(start);
+  char *end = addr + chunks * chunk_size;
+  for (; addr < end; addr += chunk_size) {
+    CUDA_CALL(cuFileBufRegister(addr, chunk_size, 0));
+  }
+}
+
+void UnregisterChunks(void *start, int chunks, size_t chunk_size) {
+  char *addr = static_cast<char *>(start);
+  char *end = addr + chunks * chunk_size;
+  for (; addr < end; addr += chunk_size) {
+    CUDA_CALL(cuFileBufDeregister(addr));
+  }
+}
+
+///////////////////////////////////////////////////////////////////////////////
+// GDSRegisteredResource
+
+class GDSRegisteredResource : public mm::memory_resource<mm::memory_kind::device> {
+  void adjust_params(size_t &size, size_t &alignment) {
+    if (alignment < chunk_size_) {
```

Comment:
Is this condition sufficient? Maybe the alignment should be just a multiply of chunk_size_?

Reply:
`alignment` must be a power of 2 (and so does chunk_size), so if it's not less, then it's a multiple.

Reply:
I would add a comment here about it. It is hard to infer this silent dependency between values without a deep understanding of the internals.

Reply:
Alignments being powers of 2 is rather a common knowledge.

Reply:
Done.